daffodil-core/src/main/scala/org/apache/daffodil/dsom/walker/AbstractDSOMWalker.scala

```diff
@@ -0,0 +1,217 @@
+   */
+  protected def onComplexBegin(complexElement: ComplexTypeBase): Unit
+  /**
+   * Method to be called after a "complex" element has been processed.
+   * @param complexElement the complex element that was encountered
+   */
+  protected def onComplexEnd(complexElement: ComplexTypeBase): Unit
+
+  /**
+   * Method to be called when a regular element is encountered.  This is one of the most
+   * important methods; most DFDL elements with useful data become some subclass of
+   * ElementBase.
+   * @param element the element that was encountered
+   */
+  protected def onElementBegin(element: ElementBase): Unit
+  /**
+   * Method to be called when a regular element has been processed.
+   * @param element the element that was encountered
+   */
+  protected def onElementEnd(element: ElementBase): Unit
+
+  final def walkDSOMSchema(): T = {
+    onWalkBegin(schemaSetRoot)
+    // this is allowed because Root is a subclass of ElementBase
+    walkerHelper(schemaSetRoot)
+    // onWalkEnd will return the value returned by this method!
+    onWalkEnd(schemaSetRoot)
+  }
+
+  /**
+   * Overloaded walkerHelper method used for when some "group" element is encountered in
+   * the DSOM; the group elements are Sequences, Choices, and GroupRefs
+   * @param modelGroup the ModelGroup element that was encountered in the Schema
+   */
+  private final def walkerHelper(modelGroup: ModelGroup): Unit = {
+    modelGroup match {
+      case sequence: Sequence =>
+        onSequenceBegin(sequence)
+        for (child <- sequence.groupMembers) walkerHelper(child)
+        onSequenceEnd(sequence)
+      case choice: Choice =>
+        onChoiceBegin(choice)
+        for (child <- choice.groupMembers) walkerHelper(child)
+        onChoiceEnd(choice)
+      case groupRef: GroupRef =>
+        if (!groupRef.isHidden) {
+          onGroupRefBegin(groupRef)
+          for (child <- groupRef.groupMembers) walkerHelper(child)
+          onGroupRefEnd(groupRef)
+        }
+    }
+  }
+
+  /**
+   * Overloaded walkerHelper method that will handle the input either as an ElementBase or ModelGroup
+   * @param term the Term element that was encountered in the Schema
+   */
+  private final def walkerHelper(term: Term): Unit = {
+    term match {
+      case element: ElementBase => walkerHelper(element)
+      case modelGroup: ModelGroup => walkerHelper(modelGroup)
+    }
+  }
+
+  /**
+   * Overloaded walkerHelper method that will handle some element either as a complex or simple type
+   * @param element the element that was encountered in the Schema
+   */
+  private final def walkerHelper(element: ElementBase): Unit = {
+    onElementBegin(element)
+    if (element.isComplexType) {
+      onComplexBegin(element.complexType)
+      walkerHelper(element.complexType.group)
+      onComplexEnd(element.complexType)
+    } else {
+      onSimpleBegin(element.simpleType)
+      onSimpleEnd(element.simpleType)
+    }
+    onElementEnd(element)
+  }
+}
```

Comment:
I'm not sure much is gained by spliting the walkerHelper's into different overloaded methods. Just sort of adds another layer of indirection. I think I'd rather see it as one function that accepts a Term and is something like:
```scala
term match {
 case s: Sequence => ...
 case c: Choice => ...
 case g: GroupRef => ...
 case: e: ElementBase => ...
}
```
Another option, taking my first comment a  little further, perhaps instead of the walker having the logic for walking, it all moves into the DSOM. We require that each SchemaComponent implement a ``def walkDSOM(w: Walker)`` method. So for example:
```scala
class Seqeunce {
  ...
  override def walkDSOM(w: Walker): Unit = {
    w.onSequenceBegin(...)
    groupMembers.forEach(_.walkDSOM(w))
    w.onSequenceEnd(...)
  }
```
It's sortof two sides of the same coin, but this way the DSOM elements are more in reponsible for knowing how to walk their own DSOM rather than the walker having to know how to walk everything DSOM. And the AbstractWalker essentialy just becomes a WalkerHandler of events, and doesn't do any actual walking itself.

Reply:
Sometimes nothing but a recursive walk algorithm will do. A visitor struggles with state management when it has to refer to more than one tree node at a time. 

Suggest we stick for now with what makes this particular NiFI Record-Schema synthesizer work out well. If that's a visit pattern then great, but if an actual walk proves more natural we should have that.

Reply:
While I did leave the fundamental recursive walk structure intact, as Mike stated might have to do anyway in the end, I did consolidate quite a few methods.  Now the user need only implement events for starting and ending the walk, and 2 handlers each for Terms and Types, where a Type could either be Simple or Complex.  See the latest commit.